File: backend/app/api/routes/cti.py
"""
UMEagleEye - CTI API routes for threat indicators and MITRE ATT&CK matrix.
"""

from typing import Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import select, func

from app.core.dependencies import get_db, get_current_user
from app.db.models import CTIIndicator, User
from app.schemas.models import CTIIndicatorResponse, CTIListResponse

router = APIRouter(prefix="/cti", tags=["Threat Intelligence"])


@router.get("/indicators", response_model=CTIListResponse)
async def list_indicators(
    source: Optional[str] = None,
    indicator_type: Optional[str] = None,
    limit: int = Query(100, ge=1, le=500),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """List CTI indicators with optional filtering."""
    query = select(CTIIndicator)

    if source:
        query = query.where(CTIIndicator.source.ilike(f"%{source}%"))
    if indicator_type:
        query = query.where(CTIIndicator.indicator_type == indicator_type)

    count_query = select(func.count()).select_from(query.subquery())
    total = (await db.execute(count_query)).scalar()

    query = query.order_by(CTIIndicator.last_seen.desc()).limit(limit)
    result = await db.execute(query)
    indicators = result.scalars().all()

    return CTIListResponse(
        items=[CTIIndicatorResponse.model_validate(i) for i in indicators],
        total=total,
    )


@router.get("/matrix")
async def get_attack_matrix(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get MITRE ATT&CK matrix data from stored indicators."""
    from app.services.threat_intel import ThreatIntelService

    result = await db.execute(select(CTIIndicator).limit(1000))
    indicators = result.scalars().all()

    indicator_dicts = [
        {
            "attack_tactic": i.attack_tactic,
            "attack_technique": i.attack_technique,
            "source": i.source,
        }
        for i in indicators
    ]

    matrix = ThreatIntelService.get_attack_matrix_data(indicator_dicts)
    return {"matrix": matrix, "total_indicators": len(indicators)}


@router.post("/ingest")
async def trigger_cti_ingestion(
    current_user: User = Depends(get_current_user),
):
    """Manually trigger CTI feed ingestion."""
    from app.tasks.cti_tasks import ingest_mycert_feeds
    task = ingest_mycert_feeds.delay()
    return {"task_id": task.id, "status": "QUEUED"}


@router.get("/lookup")
async def lookup_indicator(
    value: str = Query(..., description="IP, domain, URL, or hash to look up"),
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Look up a specific IOC value in the threat intelligence database.

    Returns the indicator record if found, plus a cross-reference
    against internal assets for immediate triage context.
    """
    from app.db.models import Asset

    # Find the indicator
    result = await db.execute(
        select(CTIIndicator).where(CTIIndicator.value == value.strip())
    )
    indicator = result.scalar_one_or_none()

    # Cross-reference against assets (IP only)
    matched_asset = None
    try:
        asset_result = await db.execute(
            select(Asset).where(Asset.ip_address == value.strip()).limit(1)
        )
        asset = asset_result.scalar_one_or_none()
        if asset:
            matched_asset = {
                "asset_id": str(asset.asset_id),
                "hostname": asset.hostname,
                "ip_address": str(asset.ip_address),
                "criticality_score": asset.criticality_score,
                "is_internet_facing": asset.is_internet_facing,
            }
    except Exception:
        pass  # Non-IP lookups will fail the INET cast — expected

    if indicator:
        return {
            "found": True,
            "indicator": {
                "indicator_id": str(indicator.indicator_id),
                "source": indicator.source,
                "indicator_type": indicator.indicator_type.value,
                "value": indicator.value,
                "confidence_score": float(indicator.confidence_score) if indicator.confidence_score else None,
                "attack_tactic": indicator.attack_tactic,
                "attack_technique": indicator.attack_technique,
                "first_seen": indicator.first_seen.isoformat(),
                "last_seen": indicator.last_seen.isoformat(),
            },
            "matched_internal_asset": matched_asset,
        }

    return {
        "found": False,
        "value": value,
        "matched_internal_asset": matched_asset,
    }


@router.get("/stats")
async def get_cti_stats(
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    """Get CTI statistics summary."""
    total = (await db.execute(select(func.count(CTIIndicator.indicator_id)))).scalar()

    # Count by source
    sources = (await db.execute(
        select(CTIIndicator.source, func.count(CTIIndicator.indicator_id))
        .group_by(CTIIndicator.source)
    )).all()

    # Count by type
    types = (await db.execute(
        select(CTIIndicator.indicator_type, func.count(CTIIndicator.indicator_id))
        .group_by(CTIIndicator.indicator_type)
    )).all()

    return {
        "total_indicators": total,
        "by_source": {s: c for s, c in sources},
        "by_type": {t.value if hasattr(t, 'value') else t: c for t, c in types},
    }

File: workers/src/routes/cti.ts
/**
 * UMEagleEye Workers — CTI routes
 * Returns snake_case field names to match frontend expectations.
 */

import { Hono } from 'hono'
import { eq, and, desc, gte, ilike, sql } from 'drizzle-orm'
import type { Env } from '../types'
import { authMiddleware, requireRoles } from '../middleware/auth'
import { getDb } from '../db/client'
import { ctiIndicators, assets } from '../db/schema'

const app = new Hono<{ Bindings: Env }>()

const INGEST_ROLES = ['ops_lead', 'security_engineer', 'superadmin']

// ── Helper: transform a Drizzle CTI row → snake_case ─────────────
function toSnake(row: typeof ctiIndicators.$inferSelect) {
  return {
    indicator_id:       row.indicatorId,
    source:             row.source,
    indicator_type:     row.indicatorType,
    value:              row.value,
    confidence_score:   row.confidenceScore != null ? Number(row.confidenceScore) : null,
    attack_tactic:      row.attackTactic,
    attack_technique:   row.attackTechnique,
    first_seen:         row.firstSeen,
    last_seen:          row.lastSeen,
  }
}

// ── GET /indicators ──────────────────────────────────────────────
app.get('/indicators', authMiddleware, async (c) => {
  try {
    const db = getDb(c.env.DATABASE_URL)

    const page     = Math.max(1, parseInt(c.req.query('page')  ?? '1'))
    const limit    = Math.min(500, Math.max(1, parseInt(c.req.query('limit') ?? '100')))
    const offset   = (page - 1) * limit
    const type     = c.req.query('indicator_type') ?? c.req.query('type')
    const source   = c.req.query('source')
    const search   = c.req.query('search')

    const conditions = []
    if (type)   conditions.push(eq(ctiIndicators.indicatorType, type as 'ip' | 'domain' | 'hash' | 'url' | 'email'))
    if (source) conditions.push(ilike(ctiIndicators.source, `%${source}%`))
    if (search) conditions.push(ilike(ctiIndicators.value, `%${search}%`))

    const whereClause = conditions.length > 0 ? and(...conditions) : undefined

    const [rows, countRows] = await Promise.all([
      db.select().from(ctiIndicators).where(whereClause).orderBy(desc(ctiIndicators.lastSeen)).limit(limit).offset(offset),
      db.select({ count: sql<number>`count(*)::int` }).from(ctiIndicators).where(whereClause),
    ])

    return c.json({
      total:  countRows[0]?.count ?? 0,
      page,
      limit,
      items:  rows.map(toSnake),
    })
  } catch (err) {
    console.error('cti GET /indicators error:', err)
    return c.json({ detail: 'Failed to fetch CTI indicators' }, 500)
  }
})

// ── GET /stats ───────────────────────────────────────────────────
app.get('/stats', authMiddleware, async (c) => {
  try {
    const db = getDb(c.env.DATABASE_URL)
    const sevenDaysAgo = new Date(Date.now() - 7 * 24 * 60 * 60 * 1000)

    const [totalResult, bySource, byType, recent7d] = await Promise.all([
      db.select({ count: sql<number>`count(*)::int` }).from(ctiIndicators),
      db
        .select({ source: ctiIndicators.source, count: sql<number>`count(*)::int` })
        .from(ctiIndicators)
        .groupBy(ctiIndicators.source)
        .orderBy(desc(sql`count(*)`)),
      db
        .select({ indicator_type: ctiIndicators.indicatorType, count: sql<number>`count(*)::int` })
        .from(ctiIndicators)
        .groupBy(ctiIndicators.indicatorType)
        .orderBy(desc(sql`count(*)`)),
      db
        .select({ count: sql<number>`count(*)::int` })
        .from(ctiIndicators)
        .where(gte(ctiIndicators.firstSeen, sevenDaysAgo)),
    ])

    const by_source: Record<string, number> = {}
    for (const row of bySource) by_source[row.source] = row.count

    const by_type: Record<string, number> = {}
    for (const row of byType) by_type[row.indicator_type] = row.count

    return c.json({
      total_indicators: totalResult[0]?.count ?? 0,  // snake_case for frontend
      by_source,
      by_type,
      recent_7d: recent7d[0]?.count ?? 0,
    })
  } catch (err) {
    console.error('cti GET /stats error:', err)
    return c.json({ detail: 'Failed to fetch CTI stats' }, 500)
  }
})

// ── GET /matrix ──────────────────────────────────────────────────
// Returns { matrix: { "Tactic Name": [{technique_id, count}] } }
app.get('/matrix', authMiddleware, async (c) => {
  try {
    const db = getDb(c.env.DATABASE_URL)

    const rows = await db
      .select({
        attack_tactic:    ctiIndicators.attackTactic,
        attack_technique: ctiIndicators.attackTechnique,
        count:            sql<number>`count(*)::int`,
      })
      .from(ctiIndicators)
      .where(sql`${ctiIndicators.attackTactic} IS NOT NULL`)
      .groupBy(ctiIndicators.attackTactic, ctiIndicators.attackTechnique)
      .orderBy(ctiIndicators.attackTactic, desc(sql`count(*)`))

    // Build { tactic → [{technique_id, count}] } structure
    const matrix: Record<string, Array<{ technique_id: string; count: number }>> = {}
    for (const row of rows) {
      const tactic = row.attack_tactic!
      if (!matrix[tactic]) matrix[tactic] = []
      matrix[tactic].push({
        technique_id: row.attack_technique || 'Unknown',
        count: row.count,
      })
    }

    return c.json({ matrix, total_indicators: rows.reduce((s, r) => s + r.count, 0) })
  } catch (err) {
    console.error('cti GET /matrix error:', err)
    return c.json({ detail: 'Failed to fetch MITRE ATT&CK matrix' }, 500)
  }
})

// ── GET /lookup ──────────────────────────────────────────────────
app.get('/lookup', authMiddleware, async (c) => {
  try {
    const db    = getDb(c.env.DATABASE_URL)
    const value = c.req.query('value')?.trim()
    if (!value) return c.json({ detail: 'value query param required' }, 422)

    // Find indicator
    const [indicator] = await db
      .select()
      .from(ctiIndicators)
      .where(eq(ctiIndicators.value, value))
      .limit(1)

    // Cross-reference against internal assets (IP match only)
    let matched_internal_asset = null
    try {
      const [asset] = await db
        .select({
          assetId:         assets.assetId,
          hostname:        assets.hostname,
          ipAddress:       assets.ipAddress,
          criticalityScore: assets.criticalityScore,
          isInternetFacing: assets.isInternetFacing,
        })
        .from(assets)
        .where(eq(assets.ipAddress, value))
        .limit(1)

      if (asset) {
        matched_internal_asset = {
          asset_id:         asset.assetId,
          hostname:         asset.hostname,
          ip_address:       asset.ipAddress,
          criticality_score: asset.criticalityScore,
          is_internet_facing: asset.isInternetFacing,
        }
      }
    } catch {
      // Non-IP values will not match asset IPs — expected
    }

    if (indicator) {
      return c.json({
        found: true,
        indicator: toSnake(indicator),
        matched_internal_asset,
      })
    }

    return c.json({ found: false, value, matched_internal_asset })
  } catch (err) {
    console.error('cti GET /lookup error:', err)
    return c.json({ detail: 'Lookup failed' }, 500)
  }
})

// ── GET /debug ───────────────────────────────────────────────────
// Live reachability test — accessible to any authenticated user.
app.get('/debug', authMiddleware, async (c) => {
  const results: Record<string, unknown> = {}

  // ── OTX ──
  try {
    const key = c.env.OTX_API_KEY
    results.otx_key_set    = !!key
    results.otx_key_length = key?.length ?? 0
    const res = await fetch(
      'https://otx.alienvault.com/api/v1/pulses/activity?limit=3&page=1',
      { headers: key ? { 'X-OTX-API-KEY': key } : {} },
    )
    results.otx_http_status = res.status
    results.otx_ok = res.ok
    if (res.ok) {
      const d = await res.json() as { count?: number; results?: Array<{ name: string; indicators: unknown[] }> }
      results.otx_total_pulses_available = d.count
      results.otx_pulses_in_response     = d.results?.length ?? 0
      results.otx_first_pulse_name       = d.results?.[0]?.name ?? null
      results.otx_first_pulse_indicators = d.results?.[0]?.indicators?.length ?? 0
    } else {
      results.otx_error_body = await res.text().catch(() => '(unreadable)')
    }
  } catch (e) {
    results.otx_fetch_exception = String(e)
  }

  // ── ThreatFox ──
  try {
    const key = c.env.THREATFOX_API_KEY
    results.threatfox_key_set    = !!key
    results.threatfox_key_length = key?.length ?? 0
    const headers: Record<string, string> = { 'Content-Type': 'application/json' }
    if (key) headers['Auth-Key'] = key

    const res = await fetch('https://threatfox-api.abuse.ch/api/v1/', {
      method: 'POST',
      headers,
      body: JSON.stringify({ query: 'get_iocs', days: 1 }),
    })
    results.threatfox_http_status = res.status
    results.threatfox_ok = res.ok
    const rawText = await res.text()
    results.threatfox_raw_response_snippet = rawText.slice(0, 300)
    if (res.ok) {
      try {
        const d = JSON.parse(rawText) as { query_status?: string; data?: unknown[] }
        results.threatfox_query_status  = d.query_status
        results.threatfox_ioc_count     = Array.isArray(d.data) ? d.data.length : `data is ${typeof d.data}`
        results.threatfox_sample_ioc    = Array.isArray(d.data) ? d.data[0] : d.data
      } catch { results.threatfox_parse_error = 'JSON.parse failed' }
    }
  } catch (e) {
    results.threatfox_fetch_exception = String(e)
  }

  return c.json(results)
})

// ── POST /ingest ─────────────────────────────────────────────────
app.post('/ingest', authMiddleware, requireRoles(...INGEST_ROLES), async (c) => {
  try {
    const lockKey  = 'cti_ingest_lock'
    const existing = await c.env.KV_CACHE.get(lockKey)
    if (existing) {
      return c.json({ message: 'CTI ingestion already running or recently completed', queued: false })
    }

    await c.env.KV_CACHE.put(lockKey, '1', { expirationTtl: 3600 })
    await c.env.ADVISORY_QUEUE.send({ type: 'cti_ingest', triggeredBy: c.get('user').userId })

    return c.json({ message: 'CTI ingestion queued successfully', queued: true })
  } catch (err) {
    console.error('cti POST /ingest error:', err)
    return c.json({ detail: 'Failed to queue CTI ingestion' }, 500)
  }
})

export default app
